flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/ProcTimeWindowInnerJoin.scala

```diff
@@ -0,0 +1,326 @@
+    * @param out   The collector for returning result values.
+    *
+    */
+  override def processElement1(
+      valueC: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+
+    processElement(
+      valueC,
+      ctx,
+      out,
+      leftStreamWinSize,
+      timerState1,
+      row1MapState,
+      row2MapState,
+      -leftUpperBound,     // right stream lower
+      -leftLowerBound,     // right stream upper
+      true
+    )
+  }
+
+  /**
+    * Process right stream records
+    *
+    * @param valueC The input value.
+    * @param ctx   The ctx to register timer or get current time
+    * @param out   The collector for returning result values.
+    *
+    */
+  override def processElement2(
+      valueC: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+
+    processElement(
+      valueC,
+      ctx,
+      out,
+      rightStreamWinSize,
+      timerState2,
+      row2MapState,
+      row1MapState,
+      leftLowerBound,    // left stream upper
+      leftUpperBound,    // left stream upper
+      false
+    )
+  }
+
+  /**
+    * Called when a processing timer trigger.
+    * Expire left/right records which earlier than current time - windowsize.
+    *
+    * @param timestamp The timestamp of the firing timer.
+    * @param ctx       The ctx to register timer or get current time
+    * @param out       The collector for returning result values.
+    */
+  override def onTimer(
+      timestamp: Long,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+      out: Collector[CRow]): Unit = {
+
+    if (timerState1.value == timestamp) {
+      expireOutTimeRow(
+        timestamp,
+        leftStreamWinSize,
+        row1MapState,
+        timerState1,
+        ctx
+      )
+    }
+
+    if (timerState2.value == timestamp) {
+      expireOutTimeRow(
+        timestamp,
+        rightStreamWinSize,
+        row2MapState,
+        timerState2,
+        ctx
+      )
+    }
+  }
+
+  /**
+    * Puts an element from the input stream into state and search the other state to
+    * output records meet the condition, and registers a timer for the current record
+    * if there is no timer at present.
+    */
+  private def processElement(
+      valueC: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow],
+      winSize: Long,
+      timerState: ValueState[Long],
+      rowMapState: MapState[Long, JList[Row]],
+      oppoRowMapState: MapState[Long, JList[Row]],
+      oppoLowerBound: Long,
+      oppoUpperBound: Long,
+      isLeft: Boolean): Unit = {
+
+    cRowWrapper.out = out
+
+    val value = valueC.row
+
+    val curProcessTime = ctx.timerService.currentProcessingTime
+    val oppoLowerTime = curProcessTime + oppoLowerBound
+    val oppoUpperTime = curProcessTime + oppoUpperBound
+
+    // only when windowsize != 0, we need to store the element
+    if (winSize != 0) {
```

Comment:
I'm not sure about this.  For example, `a.proctime between b.proctime - 5 and b.proctime`. In this case, we will buffer stream `a` for a window size 5, but will not buffer stream `b` because the right window size is 0. 

Suppose the input elements are  [a1, 1], [a2, 2], [b1, 5], [a3, 5]. The first field in the tuple indicates which stream it belongs to. The seconds field in the tuple is the processing timestamp. The expected result should be `a1, b1`, `a2, b1`, `a3, b1`. But the actual result misses `a3, b1`. Because we didn't buffer the elements  from `b` stream. 

So I think, even if the window size is 0, we still need to store the elements. Of course, we will register a `curTime +1` timer to clean the states.